doc/api/esm.md

```diff
@@ -1024,6 +1024,7 @@ this one, or `undefined` if this is the main entry point for the application.
  * @param {string} specifier
  * @param {object} context
  * @param {string} context.parentURL
+ * @param {string[]} context.conditions
```

Comment:
Sorry if I’m being dense, but I still don’t understand what this PR aims to achieve; _and I wrote much of the code for hooks,_ so I feel like I really should be getting this. At the very least, add some text to the documentation explaining what `context.conditions` is and what it can be used for.

Reply:
The chain is:

1. Resolution starts within a specific context. The context includes who the parent is and certain meta information/tags about the resolution (the conditions). This includes "this is a resolution within node.js" (`node`) and "the resolution was triggered by an `import` statement or expression" (`import`).
2. The module loader calls the custom `resolve` hook with that information. It also passes the native resolution logic that does support conditional resolution as `defaultResolve` to the hook.
3. The custom `resolve` hook may inspect or modify the conditions. In the test that comes with this PR, it adds an additional condition for example (`custom-condition`). A more realistic hook may add `production` if it sees that `process.env.NODE_ENV` is set to `production`.
4. The custom `resolve` hook can call the `defaultResolve` with those modified conditions. In the test, that's exactly what it does.
5. `defaultResolve` then calls into the native C++ logic, passing the set of conditions that apply to the current resolution as provided by the custom hook.
6. When the C++ logic encounters a conditional export, it uses the set of active conditions to filter which conditions are matched. In the case of the included test, normally the C++ logic wouldn't match things guarded with `custom-condition`. But because the hook added that condition, it does now.

Reply:
Will add a brief use to the example to give some context.